File: code/python/85.py
class Solution:
    def calArea(self,heights_):
        res = 0
        heights = heights_.copy()
        heights.append(0)
        stack = []
        i = 0
        #利用单调栈，计算每个位置可以做成的最大的矩形
        while i<len(heights):
            if len(stack)==0 or heights[i]>heights[stack[-1]]:
                stack.append(i)
                i+=1
            else:
                cur = stack[-1]
                stack.pop()
                res = max(res,heights[cur]*(i if len(stack)==0 else i-stack[-1]-1))
        return res

    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        if len(matrix)==0:
            return 0
        res = 0
        heights = [0]*len(matrix[0])
        for i in range(len(matrix)):
            for j in range(len(matrix[i])):
                if matrix[i][j]=='0':
                    heights[j] = 0
                else:
                    heights[j]+=1
            #到当前层为止，各个为止的最大高度
            print(heights)
            res = max(res,self.calArea(heights))
        return res

File: code/java/85.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
        if(matrix.length==0) return 0;
        int res = 0;
        int[] heights = new int[matrix[0].length];
        int row = matrix.length;
        int col = matrix[0].length;
        for(int i=0;i<row;i++){
            for(int j=0;j<col;j++){
                if(matrix[i][j]=='0') heights[j]=0;
                else heights[j]+=1;
            }
            res = Math.max(res,calArea(heights));
        }
        return res;
    }

    public int calArea(int[] heights){
        int res = 0;
        int[] heights_ = new int[heights.length+1];
        for(int i=0;i<heights.length;i++) heights_[i] = heights[i];
        Stack stack = new Stack();
        int index = 0;
        while(index<heights_.length){
            if(stack.empty()||heights_[index]>heights_[(int)stack.peek()]){
                stack.push(index);
                index++;
            }
            else{
                int cur = (int)stack.pop();
                res = Math.max(
                    res,
                    heights_[cur]*(stack.size()==0?index:index-(int)stack.peek()-1)
                );
            }
        }
        return res;
    }
}